tools/oeedger8r/Emitter.ml

```diff
@@ -141,43 +106,48 @@ let open_file (filename : string) (dir : string) =
   fprintf os " */\n" ;
   os
 
-(** [oe_mk_struct_decl] constructs the string of a [struct] definition. *)
-let oe_mk_struct_decl (fs : string) (name : string) =
-  String.concat "\n"
-    [ sprintf "typedef struct _%s" name
-    ; "{"
-    ; "    oe_result_t _result;"
-    ; sprintf "%s} %s;" fs name ]
-
 (** [oe_gen_marshal_struct_impl] generates a marshalling [struct]
     definition. *)
-let oe_gen_marshal_struct_impl (fd : func_decl) (errno : string)
-    (isecall : bool) =
-  (* TODO: Clean this up. *)
-  let member_list_str =
-    errno
-    ^
-    let new_param_list = List.map conv_array_to_ptr fd.plist in
-    List.fold_left
-      (fun acc (pt, declr) -> acc ^ mk_ms_member_decl pt declr isecall)
-      "" new_param_list
+let oe_gen_marshal_struct =
+  let gen_member_decl (ptype : parameter_type) (decl : declarator) =
+    let tystr = get_tystr (get_param_atype ptype) in
+    let tystr =
+      if is_foreign_array ptype then
+        sprintf "/* foreign array of type %s */ void*" tystr
+      else tystr
+    in
+    (* TODO: We shouldn't have to emit a strlen variable for ocalls,
+       but currently we require it. *)
+    let need_strlen p =
+      (is_str_ptr p || is_wstr_ptr p) && (is_in_ptr p || is_inout_ptr p)
+    in
+    [ sprintf "%s %s;" tystr decl.identifier
+    ; ( if need_strlen ptype then sprintf "size_t %s_len;" decl.identifier
```

Comment:
It would be preferable to use _uint64_t_ instead of _size_t_ for variables we introduce.

Reply:
Why's that?

Reply:
We are using `size_t` in a lot of other places in the generated code already...should we address all of their uses? It looks like it's not as consistent a type as it probably should have been.

That said, this is used to compare against the [sizeof operator](https://en.cppreference.com/w/c/language/sizeof) which the standard states explicitly returns a `size_t`.

Reply:
size_t is correct is many cases.  It's just not correct any time you need cross-machine interop, such as defining fields that are transmitted between machines, or stored in files that may be read by different machines.